src/lib/transition_frontier/README.md

```diff
@@ -0,0 +1,156 @@
+# Transition Frontier
+
+The transition frontier is a hybrid in-memory/on-disk data structure which represents all known states on the network up to the point of finality. This data structure plays an essential role in tracking states produced by consensus protocol in a way that automatically garbage collects orphaned states which will never become part of the canonical chain, and does so in a way that optimizes for high read/write performance while also persisting information in the background for future recovery. As such, the transition frontier can be thought of as both an in-memory data structure, and a concurrent subsystem which maintains a persistent on-disk copy of the datastructure. To help maintain this viewpoint, the implementation is split into 2 main parts: a full frontier, which stores the entire expanded state of each block in-memory, and a persistent frontier, which asynchronously processes state transitions to the full frontier, proxying those operations to a RocksDB representation of the frontier in the background.
+
+In terms of the data structure, the transition frontier can be thought of as a combination of the following pieces of information:
+
+1) A rose tree that contains all blockchains (including forks) up to `k` in length from the most recently finalized block.
+2) A history of recently finalized blocks.
+3) A snarked ledger for the most recently finalized block.
+4) A series of ledger masks, chained off of the aforementioned snarked ledger, to represent intermediate staged ledger states achieved by blocks tracked past the most recently finalized block.
+5) The auxiliary scan state information associated with each block tracked past the most recently finalized block.
+
+Importantly, the transition frontier also can identify which of the states it is tracking is the strongest state, which is referred to as the "best tip". The consensus mechanism informs the transition frontier of how to compare blocks for strength.
+
+## Formal Spec
+
+TODO
```

Comment:
I think this will end up being a _link_ to the specs folder at the top-level so the spec is all in one place, right?